tests/core/test_lightning_optimizer.py

```diff
@@ -433,3 +435,175 @@ def configure_optimizers(self):
+                    else:
+                        optimizer.step()
+
+                    optimizer.zero_grad()
+                    if not mock:
+                        assert torch.abs(self.layer.weight.grad).sum() == 0
+
+            model = TestModel("SGD")
+            initial_weights["override"] = model.layer.weight.clone()
+        else:
+            expected_global_step = (expected_batches) // accumulate_grad_batches
+            if enable_pl_optimizer:
+                model = TestBoringModel("Adam" if mock else "SGD")
+                initial_weights["after"] = model.layer.weight.clone()
+            else:
+                model = TestBoringModel("AdamW" if mock else "SGD")
+                initial_weights["before"] = model.layer.weight.clone()
+
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            max_epochs=max_epochs,
+            limit_train_batches=limit_train_batches,
+            limit_val_batches=limit_val_batches,
+            enable_pl_optimizer=enable_pl_optimizer,
+            accumulate_grad_batches=accumulate_grad_batches if not override else 1,
+            amp_backend=amp_backend,
+            precision=precision,
+            gpus=1
+        )
+        trainer.fit(model)
+        if is_divisible:
+            assert trainer.global_step == expected_global_step
+        else:
+            assert trainer.global_step == expected_global_step + 1
+        return model
+
+    before = train(False)
+    after = train(True)
+
+    assert torch.equal(initial_weights["before"], initial_weights["after"])
+    assert len(before.losses) == expected_batches
+
+    if is_divisible:
+        assert len(before.grads) == (expected_batches // accumulate_grad_batches)
+    else:
+        assert len(before.grads) == (expected_batches // accumulate_grad_batches) + 1
+
+    assert not torch.FloatTensor(before.losses).isnan().any()
+
+    assert torch.equal(torch.FloatTensor(before.losses), torch.FloatTensor(after.losses))
+    assert before.on_before_zero_grad_count == after.on_before_zero_grad_count
+
+    for b_grad, a_grad in zip(before.grads, after.grads):
+        assert torch.equal(b_grad, a_grad), 'Grad parameters are different'
+
+    for b_w, a_w in zip(before.parameters(), after.parameters()):
+        assert torch.equal(b_w, a_w), 'Model parameters are different'
+
+    override = train(False, override=True)
+
+    assert torch.equal(initial_weights["before"], initial_weights["override"])
+    assert override.grad_checked
+    assert override.losses == before.losses
+    assert (override.on_before_zero_grad_count // accumulate_grad_batches) == before.on_before_zero_grad_count
+
+    for b_grad, o_grad in zip(before.grads, override.grads):
+        assert torch.equal(b_grad, o_grad), 'Grad parameters are different'
+
+    for b_w, o_w in zip(before.parameters(), override.parameters()):
+        assert torch.equal(b_w, o_w), 'Model parameters are different'
+
+    if precision == 32:
```

Comment:
is there a version for precision 16

Reply:
Not yet, in precision 16 with AMP native, when mocking step, we call scaler.step(optimizer). However, the step function being mocked, it doesn't get called internally and the count is 0.